src/gpuarray_blas_opencl_clblas.c

```diff
@@ -146,54 +147,145 @@ static int dgemmBatch(cb_order order, cb_transpose transA, cb_transpose transB,
-                      gpudata **x, size_t *offX, size_t incX,
-                      float beta, gpudata **y, size_t *offY, size_t incY,
+                      gpudata **x, size_t *offX, int incX,
+                      float beta, gpudata **y, size_t *offY, int incY,
                       size_t batchCount, int flags) {
   return GA_DEVSUP_ERROR;
 }
 
 static int dgemvBatch(cb_order order, cb_transpose transA,
                       size_t M, size_t N, double alpha,
                       gpudata **A, size_t *offA, size_t lda,
-                      gpudata **x, size_t *offX, size_t incX,
-                      double beta, gpudata **y, size_t *offY, size_t incY,
+                      gpudata **x, size_t *offX, int incX,
+                      double beta, gpudata **y, size_t *offY, int incY,
                       size_t batchCount, int flags) {
   return GA_DEVSUP_ERROR;
 }
 
 static int hgerBatch(cb_order order, size_t M, size_t N, float alpha,
-                     gpudata **x, size_t *offX, size_t incX,
-                     gpudata **y, size_t *offY, size_t incY,
+                     gpudata **x, size_t *offX, int incX,
+                     gpudata **y, size_t *offY, int incY,
                      gpudata **A, size_t *offA, size_t lda,
                      size_t batchCount, int flags) {
   return GA_DEVSUP_ERROR;
 }
 
 static int sgerBatch(cb_order order, size_t M, size_t N, float alpha,
-                     gpudata **x, size_t *offX, size_t incX,
-                     gpudata **y, size_t *offY, size_t incY,
+                     gpudata **x, size_t *offX, int incX,
+                     gpudata **y, size_t *offY, int incY,
                      gpudata **A, size_t *offA, size_t lda,
                      size_t batchCount, int flags) {
   return GA_DEVSUP_ERROR;
 }
 
 static int dgerBatch(cb_order order, size_t M, size_t N, double alpha,
-                     gpudata **x, size_t *offX, size_t incX,
-                     gpudata **y, size_t *offY, size_t incY,
+                     gpudata **x, size_t *offX, int incX,
+                     gpudata **y, size_t *offY, int incY,
                      gpudata **A, size_t *offA, size_t lda,
                      size_t batchCount, int flags) {
   return GA_DEVSUP_ERROR;
 }
 
+static int hdot(
+        size_t N,
+        gpudata *X, size_t offX, int incX,
+        gpudata *Y, size_t offY, int incY,
+        gpudata *Z, size_t offZ) {
+    return GA_DEVSUP_ERROR;
+}
+
+static int sdot(
+        size_t N,
+        gpudata *X, size_t offX, int incX,
+        gpudata *Y, size_t offY, int incY,
+        gpudata *Z, size_t offZ) {
+  cl_ctx *ctx = X->ctx;
+  clblasStatus err;
+  cl_int cl_err;
+  cl_uint num_ev = 0;
+  cl_event evl[3];
+  cl_event ev;
+  cl_mem scratch_mem;
+
+  scratch_mem = clCreateBuffer(
+          ctx->ctx, CL_MEM_READ_WRITE, N*sizeof(float), NULL, &cl_err);
+  if (cl_err != CL_SUCCESS)
+      return GA_MEMORY_ERROR;
```

Comment:
The temporary buffer allocation is ok, however please use cl_alloc() to allocate it so we can have one central location for all allocations.